op-acceptor/logging/filelogger.go

```diff
@@ -720,3 +751,636 @@ func (s *ConciseSummarySink) Complete(runID string) error {
+				}
+			}
+		}
+
+		// Extract expected value
+		if strings.Contains(outputText, "expected:") {
+			parts := strings.Split(outputText, "expected:")
+			if len(parts) > 1 {
+				expectedLine := strings.TrimSpace(parts[1])
+				endIdx := strings.Index(expectedLine, "\n")
+				if endIdx > 0 {
+					info.Expected = expectedLine[:endIdx]
+				} else {
+					info.Expected = expectedLine
+				}
+			}
+		}
+
+		// Extract actual value
+		if strings.Contains(outputText, "actual") {
+			parts := strings.Split(outputText, "actual")
+			if len(parts) > 1 {
+				actualLine := strings.TrimSpace(parts[1])
+				endIdx := strings.Index(actualLine, "\n")
+				if endIdx > 0 {
+					info.Actual = actualLine[:endIdx]
+				} else {
+					info.Actual = actualLine
+				}
+			}
+		}
+
+		// Extract messages
+		if strings.Contains(outputText, "Messages:") {
+			parts := strings.Split(outputText, "Messages:")
+			if len(parts) > 1 {
+				message := strings.TrimSpace(parts[1])
+				endIdx := strings.Index(message, "\n")
+				if endIdx > 0 {
+					info.Messages += message[:endIdx] + "\n"
+				} else {
+					info.Messages += message + "\n"
+				}
+			}
+		}
+	}
+
+	return info
+}
+
+// Complete is a no-op for PerTestFileSink
+func (s *PerTestFileSink) Complete(runID string) error {
+	return nil
+}
+
+// HTMLSummarySink creates an HTML report for better visualization of test results
+type HTMLSummarySink struct {
+	logger      *FileLogger
+	passed      int
+	failed      int
+	skipped     int
+	errored     int
+	testResults []*types.TestResult
+}
+
+// Consume collects test results for later HTML generation
+func (s *HTMLSummarySink) Consume(result *types.TestResult, runID string) error {
+	// Store the test result for later processing
+	s.testResults = append(s.testResults, result)
+
+	// Update statistics based on test status
+	switch result.Status {
+	case types.TestStatusPass:
+		s.passed++
+	case types.TestStatusFail:
+		s.failed++
+	case types.TestStatusSkip:
+		s.skipped++
+	case types.TestStatusError:
+		s.errored++
+	}
+
+	return nil
+}
+
+// Complete generates the HTML summary file
+func (s *HTMLSummarySink) Complete(runID string) error {
+	// Get the base directory for this runID
+	baseDir, err := s.logger.GetDirectoryForRunID(runID)
+	if err != nil {
+		return err
+	}
+
+	// Create the HTML report filepath
+	htmlFile := filepath.Join(baseDir, "test-summary.html")
+
+	// Get or create the async writer
+	writer, err := s.logger.getAsyncWriter(htmlFile)
+	if err != nil {
+		return err
+	}
+
+	// Calculate totals and duration
+	total := s.passed + s.failed + s.skipped + s.errored
+
+	// Calculate total duration from sum of all test durations
+	var totalDuration time.Duration
+	for _, result := range s.testResults {
+		totalDuration += result.Duration
+	}
+
+	passRate := 0.0
+	if total > 0 {
+		passRate = float64(s.passed) / float64(total) * 100
+	}
+
+	// Build the HTML content
+	var html strings.Builder
+
+	// HTML header
+	html.WriteString(`<!DOCTYPE html>
```

Comment:
I would suggest to move that content into a separate page.html.tmpl file containing a proper Go template (using "html/template" at least, or even better "github.com/google/safehtml/template") as there's potentially externally-controlled data making its way to this page. So we need to make sure we protect whoever is gonna load it (generally that'd be us :D)

You can then use "embed" to manipulate it from the code

Reply:
Done (with html/template)

Reply:
Also implemented a fallback, as the test for this was failing in CI (and only in CI).